cli/lsp/resolver.rs

```diff
@@ -419,64 +429,111 @@ impl LspNpmConfigHash {
   cache: Arc<dyn HttpCache>,
-  redirects: Mutex<HashMap<ModuleSpecifier, ModuleSpecifier>>,
+  entries: DashMap<ModuleSpecifier, Option<Arc<RedirectEntry>>>,
 }
 
 impl RedirectResolver {
   pub fn new(cache: Arc<dyn HttpCache>) -> Self {
     Self {
       cache,
-      redirects: Mutex::new(HashMap::new()),
+      entries: Default::default(),
     }
   }
 
   pub fn resolve(
     &self,
     specifier: &ModuleSpecifier,
   ) -> Option<ModuleSpecifier> {
-    if matches!(specifier.scheme(), "http" | "https") {
-      let mut redirects = self.redirects.lock();
-      if let Some(specifier) = redirects.get(specifier) {
-        Some(specifier.clone())
+    if !matches!(specifier.scheme(), "http" | "https") {
+      return Some(specifier.clone());
+    }
+    let mut current = specifier.clone();
+    let mut chain = vec![];
+    let destination = loop {
+      if let Some(maybe_entry) = self.entries.get(&current) {
+        break match maybe_entry.as_ref() {
+          Some(entry) => entry.destination.clone(),
+          None => Some(current),
+        };
+      }
+      let Some(cache_key) = self.cache.cache_item_key(&current).ok() else {
+        break None;
+      };
+      let Some(headers) = self.cache.read_headers(&cache_key).ok().flatten()
+      else {
+        break None;
+      };
+      let headers = Arc::new(headers);
+      if let Some(location) = headers.get("location") {
+        if chain.len() > 10 {
+          break None;
+        }
+        let Ok(target) =
+          deno_core::resolve_import(location, specifier.as_str())
+        else {
+          break None;
+        };
+        chain.push((
+          current.clone(),
+          RedirectEntry {
+            headers,
+            target: target.clone(),
+            destination: None,
+          },
+        ));
+        current = target;
       } else {
-        let redirect = self.resolve_remote(specifier, 10)?;
-        redirects.insert(specifier.clone(), redirect.clone());
-        Some(redirect)
+        self.entries.insert(current.clone(), None);
+        break Some(current);
       }
-    } else {
-      Some(specifier.clone())
+    };
+    for (specifier, mut entry) in chain {
+      entry.destination = destination.clone();
+      self.entries.insert(specifier, Some(Arc::new(entry)));
     }
+    destination
   }
 
-  fn resolve_remote(
+  fn chain(
     &self,
     specifier: &ModuleSpecifier,
-    redirect_limit: usize,
-  ) -> Option<ModuleSpecifier> {
-    if redirect_limit > 0 {
-      let cache_key = self.cache.cache_item_key(specifier).ok()?;
-      let headers = self.cache.read_headers(&cache_key).ok().flatten()?;
-      if let Some(location) = headers.get("location") {
-        let redirect =
-          deno_core::resolve_import(location, specifier.as_str()).ok()?;
-        self.resolve_remote(&redirect, redirect_limit - 1)
-      } else {
-        Some(specifier.clone())
+  ) -> impl IntoIterator<Item = (ModuleSpecifier, Arc<RedirectEntry>)> {
```

Comment:
Looks like this could just return a `Vec<_>`?

Feel like the use of this could have the last three lines removed, but maybe I'm misreading?

```rs
redirect_resolver
      .chain(specifier)
      .into_iter()
      .map(|(s, e)| (s, e.headers.clone()))
      .collect()
```